providers/redfish/firmware.go

```diff
@@ -120,59 +124,82 @@ func (c *Conn) FirmwareInstall(ctx context.Context, component, applyAt string, f
 
 	// The response contains a location header pointing to the task URI
 	// Location: /redfish/v1/TaskService/Tasks/JID_467696020275
-	if strings.Contains(resp.Header.Get("Location"), "JID_") {
+	var location = resp.Header.Get("Location")
+	if strings.Contains(location, "JID_") {
 		taskID = strings.Split(resp.Header.Get("Location"), "JID_")[1]
+	} else if strings.Contains(location, "/Monitor") {
+		// OpenBMC returns a monitor URL in Location
+		// Location: /redfish/v1/TaskService/Tasks/12/Monitor
+		splits := strings.Split(location, "/")
+		taskID = splits[5]
+	} else {
+		return "", bmclibErrs.ErrTaskNotFound
 	}
 
 	return taskID, nil
 }
 
-// FirmwareInstallStatus returns the status of the firmware install task queued
-func (c *Conn) FirmwareInstallStatus(ctx context.Context, installVersion, component, taskID string) (state string, err error) {
-	vendor, _, err := c.DeviceVendorModel(ctx)
+func (c *Conn) multipartHTTPUpload(ctx context.Context, url, applyAt string, update io.Reader) (*http.Response, error) {
+	if url == "" {
+		return nil, fmt.Errorf("unable to execute request, no target provided")
+	}
+
+	parameters, err := json.Marshal(struct {
+		Targets            []string `json:"Targets"`
+		RedfishOpApplyTime string   `json:"@Redfish.OperationApplyTime"`
+		Oem                struct{} `json:"Oem"`
+	}{
+		[]string{},
+		applyAt,
+		struct{}{},
+	})
+
 	if err != nil {
-		return state, errors.Wrap(err, "unable to determine device vendor, model attributes")
+		return nil, errors.Wrap(err, "error preparing multipart UpdateParameters payload")
 	}
 
-	var task *gofishrf.Task
-	switch {
-	case strings.Contains(vendor, constants.Dell):
-		task, err = c.dellJobAsRedfishTask(taskID)
-	default:
-		err = errors.Wrap(
-			bmclibErrs.ErrNotImplemented,
-			"FirmwareInstallStatus() for vendor: "+vendor,
-		)
+	// payload ordered in the format it ends up in the multipart form
+	payload := []map[string]io.Reader{
+		{"UpdateParameters": bytes.NewReader(parameters)},
+		{"UpdateFile": update},
 	}
 
-	if err != nil {
-		return state, err
+	return c.runRequestWithMultipartPayload(url, payload)
+}
+
+func (c *Conn) unstructuredHttpUpload(ctx context.Context, url, applyAt string, update io.Reader) (*http.Response, error) {
+
+	if url == "" {
+		return nil, fmt.Errorf("unable to execute request, no target provided")
 	}
 
-	if task == nil {
-		return state, errors.New("failed to lookup task status for task ID: " + taskID)
+	b, _ := io.ReadAll(update)
```

Comment:
Can we try to create a goroutine that will io.Copy the contents so we don't have to buffer all the contents here?
See the `runRequestWithMultipartPayload()` method as an example

Reply:
I added a TODO because I don't really understand this for now. I don't know if I really should use a `PipeReader` or if it's related to the multipart. As this is a deprecated update mechanism, we can leave it like that, waiting for it to be removed or improved